runtime/compiler/z/codegen/J9TreeEvaluator.cpp

```diff
@@ -10363,82 +10340,66 @@ genHeapAlloc(TR::Node * node, TR::Instruction *& iCursor, bool isVariableLen, TR
-
-            // Zero the residue bytes.
+            genMemoryZeroingLoop(cg, node, iCursor, addressReg, loopCountRegister);
+
+            // Zero residue bytes.
             TR::LabelSymbol *exrlTargetLabel = generateLabelSymbol(cg);
+            iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRC, TR::InstOpCode::COND_BRC, node, ResidueBytesZeroingLabel);
             iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, exrlTargetLabel);
             iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 0, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
-            iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerRemLabel);
-            iCursor = generateRILInstruction(cg, TR::InstOpCode::EXRL, node, lengthReg, exrlTargetLabel);
+            iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, ResidueBytesZeroingLabel);
+            iCursor = generateRILInstruction(cg, TR::InstOpCode::EXRL, node, sizePreserveRegister, exrlTargetLabel);
+
+            srm->reclaimScratchRegister(addressReg);
+            srm->reclaimScratchRegister(sizePreserveRegister);
             }
          else
             {
+            int32_t loops = allocSize / 256;
             int32_t numberOfResidueBytes = allocSize % 256;
             int32_t displacement = 0;
-            // Using a BRCT loop to clean blocks of 256 to save iCache.
-            if (generateXCLoop)
+            TR::Register * addressReg = resReg;
+            if (loops > 2)
                {
-               shiftReg = cg->allocateRegister();
-               if (conditions != NULL)
-                  conditions->addPostCondition(shiftReg, TR::RealRegister::AssignAny);
-
+               // Repurpose sizeReg to keep address.
+               TR::Register * addressReg = sizeReg;
+               TR::Register * loopCountRegister = srm->findOrCreateScratchRegister();
                iCursor = generateRRInstruction(cg, TR::InstOpCode::getLoadRegOpCode(), node, addressReg, resReg, iCursor);
-               iCursor = generateLoad32BitConstant(cg, node, loops, shiftReg, true, iCursor, conditions);
-               iCursor = generateS390LabelInstruction(cg, TR::InstOpCode::label, node, fillerLoopLabel);
-               iCursor = generateSS1Instruction(cg, TR::InstOpCode::XC, node, 255, generateS390MemoryReference(addressReg, 0, cg), generateS390MemoryReference(addressReg, 0, cg), iCursor);
-               iCursor = generateRXInstruction(cg, TR::InstOpCode::LA, node, addressReg, generateS390MemoryReference(addressReg, 256, cg), iCursor);
-               iCursor = generateS390BranchInstruction(cg, TR::InstOpCode::BRCT, node, shiftReg, fillerLoopLabel);
+               iCursor = generateLoad32BitConstant(cg, node, loops, loopCountRegister, true, iCursor, conditions);
```

Comment:
When you load the loopIterCount to a register, we do not need to set the condition code (`true` parameter passed in for generateLoad32BitConstant), can you fix that?